### PR TITLE
[Send Test Mail] Use json MimeType, send data via POST method and catch json response errors

### PR DESCRIPTION
**Pull request by andrepereiradasilva**
#### Summary of Changes
- Adds the correct HTTP header mimetype for json in sendmail test response and changes the HTTP method from GET to POST so the smtp server details (host, port, user, password) doesn't appear in the server/proxies log files.
- Catch HTTP or json failure errors and display error message when that happens.
###### Correct HTTP header mimetype and HTTP method
1. Go to global config in the admin panel
2. Open the console in the chrome (F12)
3. Select "Network" tab, "XHR" separator in Chrome console
4. Click the "Send Test Mail" button. You will see a new URI in the Chrome console, click on it. Check in the "Response headers" the "Content type" HTTP header is `content-type:text/html; charset=UTF-8` and the HTTP method is `GET`
5. Apply patch
6. Repeat steps 1 (clear javascript cache Ctrl+F5) to 4, you will see in the "Response headers" the content type HTTP header is `content-type:application/json; charset=utf-8` and the HTTP method is `POST`
###### Catch HTTP or json failure errors
- With the patch previously applied.
- Add in https://github.com/andrepereiradasilva/joomla-cms/blob/patch-6/administrator/components/com_config/controller/application/sendtestmail.php#L42

``` php
trigger_error('aaa', E_USER_WARNING);
```
- Go to global config in the admin panel and send a test mail
- You will get an error message (without this patch you get no message - javascript error).
  image

### CODE REVIEW

**Comment by brianteeman** [comment]
is #5238 related?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9685">issues.joomla.org/joomla-cms/9685</a>.</sub>

**Comment by andrepereiradasilva** [comment]
i think that's unrelated to this issue.

**Comment by brianteeman** [comment]
ok - fair enough ;)

**Comment by mbabker** [comment]
Being completely honest with you, the last time I tried working with `JDocumentJson` it just felt completely broken (like it doesn't even JSON encode the data buffer for starters).  Also seems pretty stupid if it forces downloads (which it didn't in that case where I was using it in Chrome, and this was last month I was working with it).

BTW, this is just me being nit-picky for the sake of being nit-picky, but headers should be consistently set and sent through the application like https://github.com/joomla/joomla-cms/blob/staging/components/com_finder/controllers/suggestions.json.php does (the day components ever get real automated tests written for them using PHP's `header()` function will just make things act up).

**Comment by andrepereiradasilva** [comment]
ok just changed jquery javascript to handle the json object data without the need to use parseJSON (it's already fetched as a json object).

ready for testing.

**Comment by andrepereiradasilva** [comment]
@piotrmocko please try now i have added some ajax fail treatments in this PR.

**Comment by piotrmocko** [comment]
I can tell, that it would not help also, without testing, as I already struggled with it in my extension. But to be 101% sure I have just tested it and it did not help.

When there is a PHP warning in the output and JSON response is send correctly 
then there is HTTP code  200 and jQuery will run "done" closure.
"Fail" closure will run for HTTP code like 404, 500, etc.

**Comment by andrepereiradasilva** [comment]
no it also works on malformed json , just tested.

to trown a warning i added
```php
trigger_error('aaa', E_USER_WARNING);
```
in https://github.com/joomla/joomla-cms/pull/9685/files#diff-f0600c10281461980c2770b1951d453eR42

Result:
image

HTTP status 200. Page:
```
Warning: aaa in /path/to/joomla-staging/administrator/components/com_config/controller/application/sendtestmail.php on line 42
{"success":true,"message":null,"messages":{"success":["The email was sent successfully to <strong>test@test.com<\/strong> using <strong>SMTP<\/strong>. You should check that you've received the test email."]},"data":true}
```

**Comment by piotrmocko** [comment]
I will test it again

**Comment by andrepereiradasilva** [comment]
ok, don't forget to clear the browser cache

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 6d22f55f230f12fc092e90f1c6e1da48a1430da1<br /><br />Tested as described in the instructions, but with Firefox/Firebug and not Chrome console :tongue: <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9685">issues.joomla.org/joomla-cms/9685</a>.</sub>

**Comment by piotrmocko** [comment]
Finally it helped :) It looks like it has changed in jQuery in last years.
You may consider to display in the message:
(jqXHR.responseText.indexOf('<html') === -1 ? jqXHR.responseText : error)

**Comment by andrepereiradasilva** [comment]
@richard67 @piotrmocko
Added better error reporting,Test instructions updated. Please test now.

**Comment by richard67** [comment]
If you explain me what "mimified" means :smile: <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9685">issues.joomla.org/joomla-cms/9685</a>.</sub>

**Comment by andrepereiradasilva** [comment]
> If you explain me what "mimified" means :smile:

copy pasting ... errors... and lack of sleep

**Comment by richard67** [comment]
Ah, I thought maybe it has to do with a girl named "Mimi"<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9685">issues.joomla.org/joomla-cms/9685</a>.</sub>

**Comment by richard67** [comment]
@andrepereiradasilva The error message I get in the test looks as follows:

> A parse error as occured while processing the following JSON data:
Warning: aaa in /homepages/38/d391032706/htdocs/test7/administrator/components/com_config/controller/application/sendtestmail.php on line 43
{"success":true,"message":null,"messages":{"success":["The email was sent successfully to admin@richard-fath.de<\/strong> using PHP Mail<\/strong>. You should check that you've received the test email."]},"data":true}

Was that intentional, that the json part is in the message?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9685">issues.joomla.org/joomla-cms/9685</a>.</sub>

**Comment by brianteeman** [comment]
Who is mimi and who gave her a massage?

**Comment by andrepereiradasilva** [comment]
@richard67 yes it's intentional to show why the parse error.

**Comment by andrepereiradasilva** [comment]
> Who is mimi and who gave her a massage?

The question is: who got a massage from mimi ;)

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 69bb1de7a4427fe66572e23a1c8c6fc3041c0fe3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9685">issues.joomla.org/joomla-cms/9685</a>.</sub>

**Comment by andrepereiradasilva** [comment]
@piotrmocko can you test this so it can be RTC for 3.5.2 too?

**Comment by piotrmocko** [comment]
I will test it today
02.04.2016 13:37 "andrepereiradasilva" <notifications@github.com>
napisał(a):

**Comment by mikeveeckmans** [comment]
I have tested this item :white_check_mark: successfully on 69bb1de7a4427fe66572e23a1c8c6fc3041c0fe3<br /><br />TEST OK<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9685">issues.joomla.org/joomla-cms/9685</a>.</sub>

**Comment by piotrmocko** [comment]
I have tested this item :white_check_mark: successfully on 69bb1de

TEST OK

**Comment by brianteeman** [comment]
RTC - thanks<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9685">issues.joomla.org/joomla-cms/9685</a>.</sub>

**Comment by wilsonge** [comment]
Sorry to be a pain as this is already RTC but is there a reason why these javascript errors can't be made lang strings?

**Comment by andrepereiradasilva** [comment]
is it possible to merge this and them i will take care of the messages in a new PR?

Just saying this because @anibalsanchez wants to make a PR too (see https://github.com/anibalsanchez/joomla-cms/pull/1) to correct a bug.
But since this has a minified js it gets conflicts and we have to redo the js.

**Comment by andrepereiradasilva** [comment]
ok, will made the changes needed.

**Comment by andrepereiradasilva** [comment]
ok. language variables added.

File: administrator/language/en-GB/en-GB.com_config.ini [hunk 1]
```diff
@@ -229,6 +229,11 @@ COM_CONFIG_SAVE_SUCCESS="Configuration successfully saved."
 COM_CONFIG_SENDMAIL_ACTION_BUTTON="Send Test Mail"
 COM_CONFIG_SENDMAIL_BODY="This is a test mail sent using "_QQ_"%s"_QQ_". If you receive it, then your email settings are correct!"
 COM_CONFIG_SENDMAIL_ERROR="Test mail could not be sent."
+COM_CONFIG_SENDMAIL_JS_ERROR_CONNECTION_ABORT="A connection abort as occured while fetching the JSON data."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_config.ini]: HAS not AS

File: administrator/language/en-GB/en-GB.com_config.ini [hunk 2]
```diff
@@ -229,6 +229,11 @@ COM_CONFIG_SAVE_SUCCESS="Configuration successfully saved."
...
+COM_CONFIG_SENDMAIL_JS_ERROR_NO_CONTENT="No content has returned."
+COM_CONFIG_SENDMAIL_JS_ERROR_OTHER="An error as occured while fetching the JSON data: HTTP %s status code."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_config.ini]: HAS not AS

File: administrator/language/en-GB/en-GB.com_config.ini [hunk 3]
```diff
@@ -229,6 +229,11 @@ COM_CONFIG_SAVE_SUCCESS="Configuration successfully saved."
...
+COM_CONFIG_SENDMAIL_JS_ERROR_PARSE="A parse error as occured while processing the following JSON data:<br/><code style="_QQ_"color:inherit;white-space:pre;padding:0;margin:0;border:0;background:inherit;"_QQ_">%s</code>"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_config.ini]: HAS not AS

File: administrator/language/en-GB/en-GB.com_config.ini [hunk 4]
```diff
@@ -229,6 +229,11 @@ COM_CONFIG_SAVE_SUCCESS="Configuration successfully saved."
...
+COM_CONFIG_SENDMAIL_JS_ERROR_TIMEOUT="A timeout as occured while fetching the JSON data."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_config.ini]: HAS not AS

File: administrator/language/en-GB/en-GB.com_config.ini [hunk 5]
```diff
@@ -229,6 +229,11 @@ COM_CONFIG_SAVE_SUCCESS="Configuration successfully saved."
 COM_CONFIG_SENDMAIL_ACTION_BUTTON="Send Test Mail"
 COM_CONFIG_SENDMAIL_BODY="This is a test mail sent using "_QQ_"%s"_QQ_". If you receive it, then your email settings are correct!"
 COM_CONFIG_SENDMAIL_ERROR="Test mail could not be sent."
+COM_CONFIG_SENDMAIL_JS_ERROR_CONNECTION_ABORT="A connection abort as occured while fetching the JSON data."
+COM_CONFIG_SENDMAIL_JS_ERROR_NO_CONTENT="No content has returned."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_config.ini]: WAS not HAS

**Comment by brianteeman** [comment]
Removed the RTC label and made some comments on the language strings<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9685">issues.joomla.org/joomla-cms/9685</a>.</sub>

File: administrator/language/en-GB/en-GB.com_config.ini (hunk 1 above)
Comment (andrepereiradasilva) [inline administrator/language/en-GB/en-GB.com_config.ini]: all corrected

**Comment by andrepereiradasilva** [comment]
The latest change was adding javascript language vars as @wilsonge requested.

@mikeveeckmans, @richard67, @brianteeman can you retest to get this RTC again?

**Comment by wilsonge** [comment]
This looks MUCH better - thankyou!

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on f56e8c8278fea5dc0a2fd402adbfcbf29c7e6ba4<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9685">issues.joomla.org/joomla-cms/9685</a>.</sub>

**Comment by anibalsanchez** [comment]
I have tested this item :white_check_mark: successfully on f56e8c8278fea5dc0a2fd402adbfcbf29c7e6ba4<br /><br />Test OK<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9685">issues.joomla.org/joomla-cms/9685</a>.</sub>

**Comment by brianteeman** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9685">issues.joomla.org/joomla-cms/9685</a>.</sub>

**Comment by anibalsanchez** [comment]
I have updated #9768 with the code from this issue.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9685">issues.joomla.org/joomla-cms/9685</a>.</sub>

**Pull request merged**